app/services/order_service.rb

```diff
@@ -63,4 +64,14 @@ def add_order_items(order, cart_items)
   def clear_cart(cart_items)
     cart_items.delete_all
   end
+
+  def can_update_status?(new_status, current_user)
+    # Users can cancel their own orders
+    return true if new_status == "canceled" && @order.user == current_user
+
+    # Admins can update any status
+    return true if current_user.admin?
+
+    false
+  end
```

Comment:
I think this can be rewritten as simple as:
```
def can_update_status?(new_status, current_user)
  (new_status == "canceled" && @order.user == current_user) || current_user.admin?
end
```